pyzpl/__init__.py

```diff
@@ -82,46 +85,70 @@
     $
 """, re.VERBOSE | re.IGNORECASE)
 
+    
+def load_stream(bytes_stream, encoding='utf-8', emit_empty=False):
+    """for propname,value in load_stream(linelist, encoding='utf-8', emit_empty=False): ...
+
+    Arguments:
+        linelist   - an iterable that yields one ZPL line at a time ('file' is a suitable
+                     argument)
+        encoding   - character encoding. Any value suitable for str.decode(), default is 'utf-8'
+        emit_empty - whether to emit value-less nodes or not (default=false)
+
+    This is a generator function, yielding one tuple of (propname, value) for each name/value
+    property in the stream. The 'propname' is itself a tuple of components, representing the
+    hierarchical path to the property.
+
+    As a stream parser, it carries no state (other that the property's place in the hierarchy).
+    """
 
-def load_stream(bytes_stream, encoding='utf-8'):
     propname_stack = tuple()
     prev_indent_lvl = 0
     for lineno, raw_line in enumerate(bytes_stream):
         decoded_line = raw_line.decode(encoding)
         cleaned_line = decoded_line.rstrip("\r\n")
         line = cleaned_line.lstrip(" ")
         if not line:
-            continue
+            continue                      # skip blank lines
         spaces = len(cleaned_line) - len(line)
 
         if spaces % 4 != 0:
             msgfmt = "Illegal indent on line {}, must be a multiple of 4\n\t{}"
-            raise ValueError(msgfmt.format(spaces, repr(cleaned_line)))
+            raise ValueError(msgfmt.format(lineno, repr(cleaned_line)))
 
         indent_lvl = spaces // 4
         if indent_lvl <= prev_indent_lvl:
             propname_stack = propname_stack[:indent_lvl]
+            # this represents an outdent?
 
         nv_pair_match = NAME_VALUE_PAIR_RE.match(line)
         if nv_pair_match:
-            propname = nv_pair_match.group(1).strip()
-            value = nv_pair_match.group(2)
-            propnames = propname_stack + (propname,)
+            propname_stack += (nv_pair_match.group(1).strip(),)
             prev_indent_lvl = indent_lvl
-            yield propnames, value
+            yield propname_stack, nv_pair_match.group(2)
```

Comment:
This definitely fixes a bug (where the "bar" items didn't have their "thing" parent specified). Also, it is stylistically consistent with the handling of `propname_stack` below.

Reply:
Seems to be the case yes. Conceptually this assumed that a name value pair would be a leaf node and so the propname_stack could be left untouched.